qiskit/providers/aer/backends/compatibility.py

```diff
@@ -0,0 +1,151 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Compatibility classes for changes to save instruction return types.
+
+These subclasses include a `__getattr__` method that allows existing
+code which used numpy.ndarray attributes to continue functioning with
+a deprecation warning.
+
+Numpy functions that consumed these classes should already work due to
+them having an `__array__` method for implicit array conversion.
+"""
+
+import warnings
+import numpy as np
+import qiskit.quantum_info as qi
+
+
+class Statevector(qi.Statevector):
+    """Aer result backwards compatibility wrapper for qiskit Statevector."""
+
+    def __getattr__(self, attr):
+        if hasattr(self.data, attr):
+            warnings.warn(
+                "The return type of saved statevectors has been changed from"
+                " a `numpy.ndarray` to a `qiskit.quantum_info.Statevector` as"
+                "of qiskit-aer 0.10. Accessing numpy array attributes is deprecated"
+                " and will result in an error in a future release. Use the `.data`"
+                " property to access the the stored ndarray for the Statevector.",
+                DeprecationWarning, stacklevel=2)
+            return getattr(self.data, attr)
+        return super().__getattribute__(attr)
+
+    def __eq__(self, other):
+        return (
+            isinstance(self, type(self)) and
```

Comment:
`isinstance(self, type(self))` _should_ be guaranteed to be true (but this is Python, so I'm sure we could engineer a situation where it's not).  I think perhaps you meant `isinstance(other, qi.Statevector)`, because equality should ignore the type, but equally the existing `Statevector.__eq__` should still work, since this new `Statevector` is a subclass of `qi.Statevector`.

Reply:
This was supposed to be `isinstance(self, type(other))`, whoops!

The reason for overriding was that current qi classes inherit part of their check from base classes which do `isinstance(other, type(self))`, so i needed to flip it around for the subclasses to pass tests that check equality.